Utilities/rANS/include/librans/Coder.h

```diff
@@ -0,0 +1,269 @@
+  Coder() = delete;
+
+  // Initialize a rANS encoder.
+  static void encInit(State<T>* r)
+  {
+    *r = LOWER_BOUND;
+  };
+
+  // Encodes a single symbol with range start "start" and frequency "freq".
+  // All frequencies are assumed to sum to "1 << scale_bits", and the
+  // resulting bytes get written to ptr (which is updated).
+  //
+  // NOTE: With rANS, you need to encode symbols in *reverse order*, i.e. from
+  // beginning to end! Likewise, the output bytestream is written *backwards*:
+  // ptr starts pointing at the end of the output buffer and keeps decrementing.
+  static void encPut(State<T>* r, Stream_t** pptr, uint32_t start, uint32_t freq, uint32_t scale_bits)
+  {
+    // renormalize
+    State<T> x = encRenorm(*r, pptr, freq, scale_bits);
+
+    // x = C(s,x)
+    *r = ((x / freq) << scale_bits) + (x % freq) + start;
+  };
+
+  // Flushes the rANS encoder.
+  static void encFlush(State<T>* r, Stream_t** pptr)
+  {
+    T x = *r;
+    Stream_t* ptr = *pptr;
+
+    if constexpr (needs64Bit<T>()) {
+      ptr -= 2;
+      ptr[0] = static_cast<Stream_t>(x >> 0);
+      ptr[1] = static_cast<Stream_t>(x >> 32);
+    } else {
+      ptr -= 4;
+      ptr[0] = static_cast<Stream_t>(x >> 0);
+      ptr[1] = static_cast<Stream_t>(x >> 8);
+      ptr[2] = static_cast<Stream_t>(x >> 16);
+      ptr[3] = static_cast<Stream_t>(x >> 24);
+    }
+
+    *pptr = ptr;
+  };
+
+  // Initializes a rANS decoder.
+  // Unlike the encoder, the decoder works forwards as you'd expect.
+  static void decInit(State<T>* r, Stream_t** pptr)
+  {
+    T x;
+    Stream_t* ptr = *pptr;
+
+    if constexpr (needs64Bit<T>()) {
+      x = static_cast<T>(ptr[0]) << 0;
+      x |= static_cast<T>(ptr[1]) << 32;
+      ptr += 2;
+    } else {
+      x = ptr[0] << 0;
+      x |= ptr[1] << 8;
+      x |= ptr[2] << 16;
+      x |= ptr[3] << 24;
+      ptr += 4;
+    }
+
+    *pptr = ptr;
+    *r = x;
+  };
+
+  // Returns the current cumulative frequency (map it to a symbol yourself!)
+  static uint32_t decGet(State<T>* r, uint32_t scale_bits)
+  {
+    return *r & ((1u << scale_bits) - 1);
+  };
+
+  // Advances in the bit stream by "popping" a single symbol with range start
+  // "start" and frequency "freq". All frequencies are assumed to sum to "1 << scale_bits",
+  // and the resulting bytes get written to ptr (which is updated).
+  static void decAdvance(State<T>* r, Stream_t** pptr, uint32_t start, uint32_t freq, uint32_t scale_bits)
+  {
+    T mask = (1ull << scale_bits) - 1;
+
+    // s, x = D(x)
+    T x = *r;
+    x = freq * (x >> scale_bits) + (x & mask) - start;
+
+    // renormalize
+    decRenorm(&x, pptr);
+
+    *r = x;
+  };
+
+  // Encodes a given symbol. This is faster than straight RansEnc since we can do
+  // multiplications instead of a divide.
+  //
+  // See Rans32EncSymbolInit for a description of how this works.
+  static void encPutSymbol(State<T>* r, Stream_t** pptr, EncoderSymbol<T> const* sym, uint32_t scale_bits)
+  {
+#ifdef DEBUG
```

Comment:
think this can be removed, `assert` is deactivated with the `NODEBUG` flag automatically.

Reply:
Thank you for the remark, will remove the ifdefs.